### ISSUE TITLE
haskellPackages.ghcWithPackages fails to add ghc packages

### ISSUE THREAD

**Issue by wldhx** [NixOS/nixpkgs#29164]
## Issue description

GHC yielded by `nix-shell -p haskellPackages.ghcWithPackages ...` can't import the specified packages.

Interestingly, Nix does build derivations for them and include them in propagated inputs, like so:

<details>
<pre>
$ ls /nix/store/f1hwx47syxvwvkb3zjvhm692c9p2wl60-ghc-8.0.2-with-packages/nix-support/ -l
total 4
lrwxrwxrwx 1 user user 98 Jan  1  1970 propagated-native-build-inputs -> /nix/store/dijk8w2kg0y1l27bjp3imzk7aw5rhak1-stm-2.4.4.1/nix-support/propagated-native-build-inputs
</pre>
</details>

but neither does cabal resolve them or ghc-pkg list them.

### Steps to reproduce

This seems to be a regression (?) with new nixpkgs:
```
$ nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion
"18.03pre114739.d0d905668c"

$ nix-shell -p "haskellPackages.ghcWithPackages (pkgs: with pkgs; [mtl])"

[nix-shell:~]$ ghc-pkg list mtl
/nix/store/2ymysaw94j7b0qxrkxr87xp929pl77ql-ghc-8.0.2-with-packages/lib/ghc-8.0.2/package.conf.d
    (no packages)
```

As older versions work fine:
```
$ nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion
"17.09pre112229.e619ace733"

$ nix-shell -p "haskellPackages.ghcWithPackages (pkgs: with pkgs; [mtl])"

[nix-shell:~]$ ghc-pkg list mtl
/nix/store/mr7qwwmmys2z9h2dxmxrsiybf9z2s1pj-ghc-8.0.2-with-packages/lib/ghc-8.0.2/package.conf.d
    mtl-2.2.1
```

## Technical details

* System: Fedora 25
* Nix version: nix-env (Nix) 1.11.{9,14}
* Nixpkgs version: 18.03pre114739.d0d905668c

**Comment by fpletz** [comment]
cc @peti

**Comment by peti** [comment]
This issue has been fixed in 89f5d52cf44baec14a7f542a605243777b16e7a5.
